rust/datafusion/src/optimizer/boolean_comparison.rs

```diff
@@ -0,0 +1,270 @@
+                    .iter()
+                    .map(|plan| self.optimize(plan))
+                    .collect::<Result<Vec<_>>>()?;
+
+                utils::from_plan(plan, &expr, &new_inputs)
+            }
+            LogicalPlan::TableScan { .. } | LogicalPlan::EmptyRelation { .. } => {
+                Ok(plan.clone())
+            }
+        }
+    }
+
+    fn name(&self) -> &str {
+        "boolean_comparison"
+    }
+}
+
+/// Recursively transverses the logical plan.
+fn optimize_expr(e: &Expr) -> Expr {
+    match e {
+        Expr::BinaryExpr { left, op, right } => {
+            let left = optimize_expr(left);
+            let right = optimize_expr(right);
+            match op {
+                Operator::Eq => match (&left, &right) {
+                    (Expr::Literal(ScalarValue::Boolean(b)), _) => match b {
+                        Some(true) => right,
+                        Some(false) | None => Expr::Not(Box::new(right)),
+                    },
+                    (_, Expr::Literal(ScalarValue::Boolean(b))) => match b {
+                        Some(true) => left,
+                        Some(false) | None => Expr::Not(Box::new(left)),
+                    },
+                    _ => Expr::BinaryExpr {
+                        left: Box::new(left),
+                        op: Operator::Eq,
+                        right: Box::new(right),
+                    },
+                },
+                Operator::NotEq => match (&left, &right) {
+                    (Expr::Literal(ScalarValue::Boolean(b)), _) => match b {
+                        Some(false) | None => right,
+                        Some(true) => Expr::Not(Box::new(right)),
+                    },
+                    (_, Expr::Literal(ScalarValue::Boolean(b))) => match b {
+                        Some(false) | None => left,
+                        Some(true) => Expr::Not(Box::new(left)),
+                    },
+                    _ => Expr::BinaryExpr {
+                        left: Box::new(left),
+                        op: Operator::NotEq,
+                        right: Box::new(right),
+                    },
+                },
+                _ => Expr::BinaryExpr {
+                    left: Box::new(left),
+                    op: op.clone(),
+                    right: Box::new(right),
+                },
+            }
+        }
+        Expr::Not(expr) => Expr::Not(Box::new(optimize_expr(&expr))),
+        Expr::Case {
+            expr,
+            when_then_expr,
+            else_expr,
+        } => {
+            if expr.is_none() {
+                // recurse into CASE WHEN condition expressions
+                Expr::Case {
+                    expr: None,
+                    when_then_expr: when_then_expr
+                        .iter()
+                        .map(|(when, then)| (Box::new(optimize_expr(when)), then.clone()))
+                        .collect(),
+                    else_expr: else_expr.clone(),
```

Comment:
I think you could also apply `optimize_expr` to then and else_expr:

```suggestion
                        .map(|(when, then)| (Box::new(optimize_expr(when)), optimize_expr(then))
                        .collect(),
                    else_expr: optimize_expr(else_expr),
```

Reply:
I have been going back and forth on this one. So basically what I am trying to avoid is

```CASE
    WHEN true THEN (col1 = true)
    ELSE (col1 = false)
END; 
```

Being optimized into:

```CASE
    WHEN true THEN col1
    ELSE !col1
END; 
```

This is not a valid optimization when col1 column is not typed as boolean. Although we currently don't support comparison between boolean type and none boolean type in datafusion, i am expecting us to support it in the future. Am I overthinking this?

Reply:
OK, i think what I can do is to check for column type based on plan schema and skip optimization for `col = true` case when col type is not boolean. `col = false` is always safe to be optimized into `!col`. What do you think?

Reply:
That is a good point @houqp  -- I think it applies to all expressions (not just CASE statements). You are in essence worried about changing the type of an expression from boolean to something else which is a good thing to worry about!

For columns that have types other than boolean, I would expect an expression like `col1 = true` to be eventually rewritten to `CAST(col1, boolean) = true` in which case the optimization to `CAST(col1, boolean)` is correct.

Upon reading this code a bit more, I can't recall exactly when cast's (coercions) are inserted. 

I think skipping the boolean rewrite when `col` is not a boolean is the right thing to do  -- and if that type of comparison can happen we need to update the rewrite rules for `BinaryExpr`, `NotExpr`, etc

Reply:
OK, I have updated the code to apply the folding optimization to all plan nodes and expressions wherever applicable. We currently doesn't do the `CAST` rewrite at the moment, so that needs to be handled in a separate ticket/PR to get full support for applying binary operators on expressions with different types. This PR now assumes automatic casting will be added in the future.